File: sankofa_app/lib/services/user_service.dart
import '../models/user_model.dart';
import 'api_client.dart';
import 'auth_service.dart';

class UserService {
  UserService({ApiClient? apiClient, AuthService? authService})
      : _authService = authService ?? AuthService(),
        _apiClient = apiClient ?? ApiClient(authService: authService ?? AuthService());

  final AuthService _authService;
  final ApiClient _apiClient;

  Future<UserModel?> getCurrentUser({bool forceRefresh = false}) async {
    if (!forceRefresh) {
      final stored = await _authService.getStoredUser();
      if (stored != null) {
        return stored;
      }
    }

    final response = await _apiClient.get('/api/auth/me/');
    if (response is Map<String, dynamic>) {
      final user = UserModel.fromApi(response);
      await _authService.saveUser(user);
      return user;
    }
    return null;
  }

  Future<UserModel?> refreshCurrentUser() => getCurrentUser(forceRefresh: true);

  Future<void> saveUser(UserModel user) => _authService.saveUser(user);

  Future<void> updateKycStatus(String status) async {
    final user = await getCurrentUser();
    if (user == null) {
      return;
    }
    final updated = user.copyWith(kycStatus: status, updatedAt: DateTime.now());
    await _authService.saveUser(updated);
  }

  Future<void> updateWalletBalance(double newBalance, {DateTime? walletUpdatedAt, DateTime? userUpdatedAt}) async {
    final user = await getCurrentUser();
    if (user == null) {
      return;
    }
    final now = DateTime.now();
    final updated = user.copyWith(
      walletBalance: newBalance,
      walletUpdatedAt: walletUpdatedAt ?? now,
      updatedAt: userUpdatedAt ?? now,
    );
    await _authService.saveUser(updated);
  }

  Future<void> clearSession() => _authService.clearSession();
}

File: sankofa_frontend/src/services/userService.ts
/**
 * User Service
 * Mirrors the mobile app's UserService
 */

import { apiClient } from '../lib/apiClient';
import { authService } from './authService';
import type { User } from '../lib/types';

class UserService {
  /**
   * Get current user, optionally forcing a refresh from the server
   */
  async getCurrentUser(forceRefresh: boolean = false): Promise<User | null> {
    if (!forceRefresh) {
      const stored = await authService.getStoredUser();
      if (stored) {
        return stored;
      }
    }

    try {
      const response = await apiClient.get<User>('/api/auth/me/');
      if (response) {
        await authService.saveUser(response);
        return response;
      }
    } catch {
      // Return null on error
    }
    return null;
  }

  /**
   * Refresh current user from server
   */
  async refreshCurrentUser(): Promise<User | null> {
    return this.getCurrentUser(true);
  }

  /**
   * Save user to local storage
   */
  async saveUser(user: User): Promise<void> {
    return authService.saveUser(user);
  }

  /**
   * Update KYC status locally
   */
  async updateKycStatus(status: User['kycStatus']): Promise<void> {
    const user = await this.getCurrentUser();
    if (!user) {
      return;
    }
    const updated: User = {
      ...user,
      kycStatus: status,
      updatedAt: new Date().toISOString(),
    };
    await authService.saveUser(updated);
  }

  /**
   * Update wallet balance locally
   */
  async updateWalletBalance(
    newBalance: number,
    walletUpdatedAt?: string,
    userUpdatedAt?: string
  ): Promise<void> {
    const user = await this.getCurrentUser();
    if (!user) {
      return;
    }
    const now = new Date().toISOString();
    const updated: User = {
      ...user,
      walletBalance: newBalance,
      walletUpdatedAt: walletUpdatedAt || now,
      updatedAt: userUpdatedAt || now,
    };
    await authService.saveUser(updated);
  }

  /**
   * Clear session (logout)
   */
  async clearSession(): Promise<void> {
    return authService.clearSession();
  }
}

export const userService = new UserService();
